ouroboros-consensus/src/Ouroboros/Consensus/Mempool/API.hs

```diff
@@ -0,0 +1,73 @@
+{-# LANGUAGE TypeFamilies #-}
+
+module Ouroboros.Consensus.Mempool.API (
+    Mempool(..)
+  , ApplyTx(..)
+    -- * Re-exports
+  , PrevApplied(..)
+  ) where
+
+import           Control.Monad.Except
+import           Data.Sequence (Seq)
+
+import           Control.Monad.Class.MonadSTM
+
+import           Ouroboros.Consensus.Ledger.Abstract
+
+-- TODO: Maybe we should move 'PrevApplied' someplace else
+import           Ouroboros.Storage.LedgerDB.Conf (PrevApplied (..))
+
+class UpdateLedger b => ApplyTx b where
+  -- | Generalized transaction
+  --
+  -- The mempool (and, accordingly, blocks) consist of "generalized
+  -- transactions"; this could be "proper" transactions (transferring funds) but
+  -- also other kinds of things such as update proposals, delegations, etc.
+  type family GenTx b :: *
+
+  -- | Update the ledger state with a generalized transaction
+  --
+  -- If the transaction was previously applied, expensive checks that have
+  -- previously been confirmed (such as checking cryptographic signatures)
+  -- should be omitted.
+  applyTx :: PrevApplied
+          -> GenTx b
+          -> LedgerState b
+          -> Except (LedgerError b) (LedgerState b)
+
+-- | Mempool
+--
+-- The mempool is the set of transactions that should be included in the next
+-- block. In principle this is a /set/ of all the transactions that we receive
+-- from our peers. In order to avoid flooding the network with invalid
+-- transactions,  however, we only want to keep /valid/ transactions in the
+-- mempool. That raises the question: valid with respect to which ledger state?
+--
+-- We opt for a very simple answer to this: the mempool will be interpreted
+-- as a /list/ of transactions; which are validated strictly in order, starting
+-- from the current ledger state. This has a number of advantages:
+--
+-- * It's simple to implement and it's efficient. In particular, no search for
+--   a valid subset is ever required.
+-- * When producing a block, we can simply take the longest possible prefix
+--   of transactions that fits in a block.
+-- * It supports wallets that submit dependent transactions (where later
+--   transaction depends on outputs from earlier ones).
+data Mempool m blk = Mempool {
+      -- | Add a bunch of transactions (oldest to newest)
+      --
+      -- As long as we keep the mempool entirely in-memory this could live in
+      -- @STM m@; we keep it in @m@ instead to leave open the possibility of
+      -- persistence.
+      --
+      -- Returns the transactions that were not added because they could not
+      -- be validated (in undefined order).
+      addTxs :: [GenTx blk] -> m [(GenTx blk, LedgerError blk)]
+
+      -- | Get all transactions in the mempool (oldest to newest)
+      --
+      -- Guarantees that these transactions will be valid (when applied strictly
+      -- in order) with respect to a call to 'getLedgerState' run /in the same
+      -- transaction/.
```

Comment:
Are we really going to guarantee that the ledger state and the mempool will be updated together atomically? That'll require the chain db update to also update the mempool in the same STM transaction. Or are we going to make this STM action be not just read-only and have it update the sequence if it's out of date?

The weaker condition would be that the mempool can be out of date wrt the chain/ledger state, but that we at least know when that's the case (e.g. track chain head point as a proxy for ledger state of that chain).

I'm not sure we need to rely on it being continuously in sync. It might be better to have the read operation be cheap and eventually to get the changes, but not require it be in sync. Worth considering. Not immediately obvious. Depends on what the outbound protocol handler side needs.